src/bbs2gh/Services/BbsSmbArchiveDownloader.cs

```diff
@@ -0,0 +1,227 @@
+                }
+
+                bytesRead += data.Length;
+                await _fileSystemProvider.WriteAsync(targetExportArchive, data);
+
+                LogProgress(bytesRead, sourceExportArchiveSize);
+            }
+
+            return targetExportArchiveFullPath;
+        }
+        finally
+        {
+            if (sourceExportArchiveHandle != null)
+            {
+                fileStore?.CloseFile(sourceExportArchiveHandle);
+            }
+
+            fileStore?.Disconnect();
+            _smbClient.Logoff();
+            _smbClient.Disconnect();
+        }
+    }
+
+    private void LogProgress(long downloadedBytes, long? totalBytes)
+    {
+        if (DateTime.Now < _nextProgressReport)
+        {
+            return;
+        }
+
+        var totalProgressMessage = totalBytes.HasValue
+            ? $" out of {GetLogFriendlySize(totalBytes.Value)} ({GetPercentage(downloadedBytes, totalBytes.Value)})"
+            : "";
+        _log.LogInformation($"Download archive in progress, {GetLogFriendlySize(downloadedBytes)}{totalProgressMessage} completed...");
+
+        _nextProgressReport = _nextProgressReport.AddSeconds(DOWNLOAD_PROGRESS_REPORT_INTERVAL_IN_SECONDS);
+    }
+
+    private string GetPercentage(long downloadedBytes, long totalBytes)
+    {
+        if (totalBytes is 0L)
+        {
+            return "unknown%";
+        }
+
+        var percentage = (int)(downloadedBytes * 100D / totalBytes);
+        return $"{percentage}%";
+    }
+
+    private string GetLogFriendlySize(long size)
+    {
+        const int kilobyte = 1024;
+        const int megabyte = 1024 * kilobyte;
+        const int gigabyte = 1024 * megabyte;
+
+        return size switch
+        {
+            < kilobyte => $"{size:n0} bytes",
+            < megabyte => $"{size / (double)kilobyte:n0} KB",
+            < gigabyte => $"{size / (double)megabyte:n0} MB",
+            _ => $"{size / (double)gigabyte:n2} GB"
+        };
+    }
+
+    private FileStream OpenWriteTargetExportArchive(string targetExportArchiveFullPath)
+    {
+        if (_fileSystemProvider.FileExists(targetExportArchiveFullPath))
+        {
+            throw new OctoshiftCliException($"Target export archive ({targetExportArchiveFullPath}) already exists.");
+        }
+
+        _fileSystemProvider.CreateDirectory(Path.GetDirectoryName(targetExportArchiveFullPath));
+        return _fileSystemProvider.Open(targetExportArchiveFullPath, FileMode.CreateNew);
+    }
+
+    private void ConnectToHost()
+    {
+        var isConnected = IPAddress.TryParse(_host, out var ipAddress)
+            ? _smbClient.Connect(ipAddress, SMBTransportType.DirectTCPTransport)
+            : _smbClient.Connect(_host, SMBTransportType.DirectTCPTransport);
+
+        if (!isConnected)
+        {
+            throw new OctoshiftCliException($"Unable to connect to host \"{_host}\".");
```

Comment:
For an error like this, do you think we should make it clear what stage in the process we've reached (i.e. we're trying to download using SMB)? Or is that going to be obvious from the existing log output?

Reply:
I guess it's going to be obvious since it only happens in the SMB flow.

Reply:
addressed in 28ca1022e7e2599f3e374aa83643ce8dd535f1e2